test/test_core/test_context.py

```diff
@@ -0,0 +1,69 @@
+import unittest
+from pyndemic.core.context import ContextError, ContextNotFoundError, \
+    ContextRegistrationMeta, register_context, unregister_context, \
+    get_context, generate_id, search_context
+
+
+class ContextManagerTestCase(unittest.TestCase):
+    def test_register_unregister(self):
+        """A comprehensive test for registering, getting and unregistering
+        the context"""
+        mock_context = "mock_string_instead_of_an_object"
+        register_context(42, mock_context)
+
+        # register a conflicting context
+        with self.assertRaises(ContextError):
+            register_context(42, "mock_string_instead_of_an_object")
+
+        # get the existing context
+        ctx = get_context(42)
+        self.assertEqual(ctx, mock_context)
+
+        unregister_context(42)
+
+        # get a non-existing context
+        with self.assertRaises(ContextNotFoundError):
+            get_context(42)
+
+        # unregister a non-existing context
+        with self.assertRaises(ContextNotFoundError):
+            unregister_context(42)
+
+    def test_generate_id(self):
+        ctx_id_1 = generate_id()
+        self.assertIsNotNone(ctx_id_1)
+        self.assertIsInstance(ctx_id_1, str)
+        self.assertEqual(8, len(ctx_id_1))
+
+        ctx_id_2 = generate_id()
+        self.assertNotEqual(ctx_id_1, ctx_id_2)
+
+
+class MockController(metaclass=ContextRegistrationMeta,
```

Comment:
Since ContextRegistrationMeta is the code we try to test, we could wrap MockController class creation in a function. Something like:
```
def construct_controller_mock_class():
    class MockController: 
        ...

    return MockController
```
This approach helps to catch exceptions from ContextRegistrationMeta while running a test case, not while parsing the "test_context.py" source file. If such an exception appears, the test log will be more clear.

Reply:
A valid point. I concluded that existence of ContextRegistrationMeta is guaranteed by addressing ._ctx property.
I tried to include this definition into setUp(), but something went wrong.